config/runtime/server.go

```diff
@@ -36,74 +35,115 @@ import (
+func GetHostPort(hostPort string) (string, int, error) {
+	var host string
+	var port int
 
-type hosts map[string]bool
-type ports map[Port]hosts
+	h, p, err := net.SplitHostPort(hostPort)
+	if err != nil {
+		return "", -1, err
+	}
 
-type HandlerKind uint8
+	host = strings.TrimRight(h, ".")
 
-const (
-	api HandlerKind = iota
-	endpoint
-	files
-	spa
-)
+	if p == "" || p == "*" {
+		port = -1
+	} else {
+		port, err = strconv.Atoi(p)
+		if err != nil {
+			return "", -1, err
+		}
+	}
 
-type endpointMap map[*config.Endpoint]*config.API
+	return host, port, nil
+}
 
 // NewServerConfiguration sets http handler specific defaults and validates the given gateway configuration.
 // Wire up all endpoints and maps them within the returned Server.
 func NewServerConfiguration(
 	conf *config.Couper, log *logrus.Entry, memStore *cache.MemoryStore,
 ) (ServerConfiguration, error) {
-	defaultPort := conf.Settings.DefaultPort
-
 	// confCtx is created to evaluate request / response related configuration errors on start.
 	noopReq := httptest.NewRequest(http.MethodGet, "https://couper.io", nil)
 	noopResp := httptest.NewRecorder().Result()
 	noopResp.Request = noopReq
 	confCtx := conf.Context.WithClientRequest(noopReq).WithBeresps(noopResp).HCLContext()
 
-	validPortMap, hostsMap, err := validatePortHosts(conf, defaultPort)
-	if err != nil {
-		return nil, err
-	}
-
 	accessControls, err := configureAccessControls(conf, confCtx)
 	if err != nil {
 		return nil, err
 	}
 
-	serverConfiguration := make(ServerConfiguration)
-	if len(validPortMap) == 0 {
-		serverConfiguration[Port(defaultPort)] = NewMuxOptions(errors.DefaultHTML, hostsMap)
-	} else {
-		for p := range validPortMap {
-			serverConfiguration[p] = NewMuxOptions(errors.DefaultHTML, hostsMap)
-		}
-	}
-
-	endpointHandlers := make(map[*config.Endpoint]http.Handler)
+	var (
+		serverConfiguration ServerConfiguration = make(ServerConfiguration)
+		defaultPort         int                 = conf.Settings.DefaultPort
+		endpointHandlers    endpointHandler     = make(endpointHandler)
+		isHostsMandatory    bool                = len(conf.Servers) > 1
+	)
 
 	for _, srvConf := range conf.Servers {
 		serverOptions, err := server.NewServerOptions(srvConf)
 		if err != nil {
 			return nil, err
 		}
 
+		if err := validateHosts(srvConf.Name, srvConf.Hosts, isHostsMandatory); err != nil {
+			return nil, err
+		}
+
+		portsHosts, err := getPortsHostsList(srvConf.Hosts, defaultPort)
+		if err != nil {
+			return nil, err
+		}
+
+		for port, hosts := range portsHosts {
+			for host, muxOpts := range hosts {
+				if serverConfiguration[port] == nil {
+					serverConfiguration[port] = make(Hosts)
+				}
+
+				if _, ok := serverConfiguration[port][host]; ok {
+					return nil, fmt.Errorf("conflict: host %q already defined for port: %d", host, port)
+				}
+
+				serverConfiguration[port][host] = muxOpts
+				serverConfiguration[port][host].ServerName = serverOptions.ServerName
```

Comment:
As already mentioned: assign `serverOptions` instead of remapping.

Reply:
Done